specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+
+This includes the following methods:
+
+- `bridgeETH(uint32,bytes)`
+- `bridgeETHTo(address,uint32,bytes)`
+- `receive()`
+
+The following diagram shows the control flow for when a user attempts to send `ether` through the `StandardBridge`, either
+on L1 or L2.
+
+![](../static/assets/custom-token-bridge-revert.png)
+
+### CrossDomainMessenger
+
+The `CrossDomainMessenger` contracts on both L1 and L2 MUST not be used to send the native asset between domains.
+It is possible to support this in the future, but for now the feature is disabled due to the high probability of footguns.
+Using the `CrossDomainMessenger` with a custom gas token would involve modifying the `sendMessage` implementation.
+
+### L1Block
+
+The `L1Block` contract is upgraded with a permissioned function that allows it to set the `GAS_PAYING_TOKEN_SLOT`.
+The `DEPOSITOR_ACCOUNT` MUST be the only account that can call the setter function. Any L2 contract that wants
+to learn the address of the gas paying token can call the getter on the `L1Block` contract.
+
+## User Flow
+
+Users should deposit native asset by calling `depositERC20Transaction` on the `OptimismPortal` contract.
+Users must first `approve` the address of the `OptimismPortal` so that the `OptimismPortal` can use
+`transferFrom` to take ownership of the ERC20 asset. If the `transferFrom` fails, a [permit2](https://github.com/Uniswap/permit2)
+`transferFrom` is attempted as a fallback.
+
+Users should withdraw value by calling the `L2ToL1MessagePasser` directly.
+
+## Upgrade
+
+The custom gas token upgrade is not yet defined to be part of a particular network upgrade, but it will be scheduled
+as part of a future hardfork. On the network upgrade block, a set of deposit transaction based upgrade transactions
+are deterministically generated by the derivation pipeline in the following order:
+
+- L1 Attributes Transaction calling `setL1BlockValuesEcotone`
+- User deposits from L1
+- Network Upgrade Transactions
+  - L1Block deployment
+  - L2CrossDomainMessenger deployment
+  - L2StandardBridge deployment
+  - Update L1Block Proxy ERC-1967 Implementation Slot
+  - Update L2CrossDomainMessenger Proxy ERC-1967 Implementation Slot
+  - Update L2StandardBridge Proxy ERC-1967 Implementation Slot
+
+### L1Block Deployment
+
+- `from`: `0x4210000000000000000000000000000000000002`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2CrossDomainMessenger Deployment
+
+- `from`: `0x4210000000000000000000000000000000000003`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2StandardBridge Deployment
+
+- `from`: `0x4210000000000000000000000000000000000004`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L1Block Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
```

Comment:
Due to the way that the proxies work, you can bypass ownership by using `address(0)`. Otherwise the derivation pipeline would have dynamic data in it, the identity of the owner of the L2 `ProxyAdmin` would need to be stolen and used which could add race conditions and L2 RPC to the fault proof/derivation pipeline so we decided to make it call from `address(0)` so its deterministic

Reply:
Added a comment addressing this